sympy/stats/stochastic_process_types.py

```diff
@@ -0,0 +1,315 @@
+    'StateSpace'
+]
+
+def _set_converter(itr):
+    """
+    Helper function for converting list/tuple/set to Set.
+    If parameter is not an instance of list/tuple/set then
+    no operation is performed.
+
+    Returns
+    =======
+
+    Set
+        The argument converted to Set.
+
+
+    Raises
+    ======
+
+    TypeError
+        If the argument is not an instance of list/tuple/set.
+    """
+    if isinstance(itr, (list, tuple, set)):
+        itr = FiniteSet(*itr)
+    if not isinstance(itr, Set):
+        raise TypeError("%s is not an instance of list/tuple/set."%(itr))
+    return itr
+
+def _matrix_checks(matrix):
+    if not isinstance(matrix, (Matrix, MatrixSymbol, ImmutableMatrix)):
+        raise TypeError("Transition probabilities etiher should "
+                            "be a Matrix or a MatrixSymbol.")
+    if matrix.shape[0] != matrix.shape[1]:
+        raise ValueError("%s is not a square matrix"%(matrix))
+    if isinstance(matrix, Matrix):
+        matrix = ImmutableMatrix(matrix.tolist())
+    return matrix
+
+class StochasticProcess(Basic):
+    """
+    Base class for all the stochastic processes whether
+    discrete or continuous.
+
+    Parameters
+    ==========
+
+    sym: Symbol or string_types
+    state_space: Set
+        The state space of the stochastic process, by default S.Reals.
+        For discrete sets it is zero indexed.
+
+    See Also
+    ========
+
+    DiscreteTimeStochasticProcess
+    """
+
+    def __new__(cls, sym, state_space=S.Reals):
+        sym = _symbol_converter(sym)
+        state_space = _set_converter(state_space)
+        return Basic.__new__(cls, sym, state_space)
+
+    @property
+    def symbol(self):
+        return self.args[0]
+
+    @property
+    def state_space(self):
+        return self.args[1]
+
+    def __call__(self, time):
+        """
+        Overrided in ContinuousTimeStochasticProcess.
+        """
+        raise NotImplementedError("Use [] for indexing discrete time stochastic process.")
+
+    def __getitem__(self, time):
+        """
+        Overrided in DiscreteTimeStochasticProcess.
+        """
+        raise NotImplementedError("Use () for indexing continuous time stochastic process.")
+
+    def probability(self, condition):
+        raise NotImplementedError()
+
+class DiscreteTimeStochasticProcess(StochasticProcess):
+    """
+    Base class for all discrete stochastic processes.
+    """
+    def __getitem__(self, time):
+        """
+        For indexing discrete time stochastic processes.
+
+        Returns
+        =======
+
+        RandomIndexedSymbol
+        """
+        if time not in self.index_set:
+            raise IndexError("%s is not in the index set of %s"%(time, self.symbol))
+        idx_obj = Indexed(self.symbol, time)
+        pspace_obj = StochasticPSpace(self.symbol, self)
+        return RandomIndexedSymbol(idx_obj, pspace_obj)
+
+# TODO: Define ContinuousTimeStochasticProcess
+
+class TransitionMatrix(Boolean):
```

Comment:
I would suggest to choose a name that makes it clear that this is an assumption. Maybe "TransitionMatrixOf" ?

Reply:
I will change the name, no issues in that. Infact, I am thinking of extending `AssumptionKeys` in `stochastic_process_types` which will reduce the number of new classes. I will try that approach and get it done.

Reply:
Wait, what do you mean by AssumptionKeys?